students/GuyCChan/session07/html_render.py

```diff
@@ -50,20 +75,79 @@ class Head(Element):
 class OneLineTag(Element):
     pass
 
-    def render(self, out_file, **attributes):
-        open_tag.write("<{} ".format(self.tag))
-        for key, value in self.attributes.items():
-            open_tag.write("{}='{}'".format(key, value))
-        open_tag.append(">\n")
+    def render(self, out_file, cur_ind="", **kwargs):
+        open_tag, close_tag = self._open_close_tags()
+        out_file.write(cur_ind)
+        out_file.write(open_tag)
         out_file.write(self.contents[0])
-        out_file.write("</{}>\n".format(self.tag))
+        out_file.write(close_tag)
+        out_file.write("\n")
 
     def append(self, content):
+        """ Cannot add content to OneLineTag."""
         raise NotImplementedError
 
 
 class Title(OneLineTag):
     tag = "title"
 
 
+class SelfClosingTag(Element):
+
+    def __init__(self, content=None, **kwargs):
+        if content is not None:
+            raise TypeError("SelfClosingTag cannot contain any content.")
+        super().__init__(content=content, **kwargs)
+
+
+    def render(self, out_file, cur_ind="", **kwargs):
+        # Please explain the use of the underscore in the following line. I tried and tried for hours
+        # and over 2 days to get this going but failed. Is it the so called "throwaway" variable?
+        # How does it work here? Without it, tests for "hr" would always come back saying "hr" has no
+        # attribute 'open_tag'.
+        open_tag, _ = self._open_close_tags()
```

Comment:
the underscore is a "throwaway" variable -- you could give it any name.

`self._open_close_tags()` returns two values: the opening tag and the closing tag. So you need to assign it to two names, in this case, `open_tag` and `_`. 

If you assign it to one name, then it doesn't raise an error, but rather sets that one name to a tuple with both values in it, so you now have a tuple of two strings, rather than a string.

Technically, functions always return one value -- either a single value or a tuple with multiple values in it. Python's "unpacking" syntax allows you to unpack that tuple inot multiple variables.

So this code could be:
```
tags = self._open_close_tags()
open_tag = tags[0]
```

and get the same results.